### DISCUSSION TITLE
Issue with client-side requests when getting data from the server with useAsyncData

### DISCUSSION THREAD

**Discussion by ukena** [nuxt/nuxt#25601] [Questions] [answered]
Hello everybody,

Im currently setting up a headless shopify app. Within the customer account the user can access its order history at the page `/account/orders` and download invoices. The invoices are stored in an AWS S3 bucket.

To download the invoices I first though about creating an api route but I later found it simpler to get the presigned Url directly within the component. I think this approach is also more secure as no `api/` server route is exposed and simpler because I dont need to authenticate requests to the server.

Below is the code snipped within the component to get the presigned URL. The `invoiceUrl` is directly used within `<UButton ... />` of the SFC.

```vue
<script setup lang="ts">
import { getSignedUrl } from '@aws-sdk/s3-request-presigner'
import { S3Client, GetObjectCommand } from '@aws-sdk/client-s3'

const props = defineProps<{
  order: any
}>()

// authenticate with aws-sdk and generate presigned url for invoice
async function getInvoiceUrl() {
  const s3 = new S3Client({
    region: 'eu-central-1',
    credentials: {
      accessKeyId: process.env.AWS_ACCESS_KEY_ID,
      secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY
    }
  })
  const command = new GetObjectCommand({
    Bucket: 'metarmap',
    Key: 'pdf/' + props.order.node.orderNumber + '.pdf'
  })
  // get presigned url for invoice which is valid for 4 hours
  const url = await getSignedUrl(s3, command, { expiresIn: 14400 })
  console.log(url)
  return url
}

const { data: invoiceUrl } = await useAsyncData(
  props.order.node.id,
  () => getInvoiceUrl()
)
</script>
```

The main problem I have is that it only works if the `/account/orders` route is hard reloaded or by calling `reloadNuxtApp` (which is a very ugly workaround). I think thats because `process.env` is not defined on the client side. So if the user navigates from e.g. `/account` to `/account/orders` the `invoiceUrl` is empty and the error 'Error: Resolved credential object is not valid' is logged.

Here is what I have tried to solve this issue:
- useState to get a consistent state and save the `invoiceUrl` there 
- use `if (process.server)` around the `getInvoiceUrl` function to make sure it is only executed on the server side
- change the route rules to `{ ssr: true, prerender: false}` to make sure it is called from the server
- use `refreshNuxtData` to get the server regenerate the presigned url

To be honest I though `useAsyncData` would solve my issue but it didn't. No matter what I try it end up with the same result that `invoiceUrl` is empty when browing to '/account/orders'.

Do you have any idea on how I can solve this?
Thanks a lot for any help!

**Comment by undermoonn** [comment, marked as answer]
I think [standalone server component](https://nuxt.com/docs/guide/directory-structure/components#standalone-server-components) is one of the solutions.

**Reply by ukena** [reply]
I moved the code to `OrderInvoiceButton.server.vue` and this solved the issue. Thanks a lot!
